datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -85,6 +91,24 @@ fn get_valid_types(
             .iter()
             .map(|valid_type| (0..*number).map(|_| valid_type.clone()).collect())
             .collect(),
+        TypeSignature::VariadicCoerced => {
+            let new_type = current_types.iter().skip(1).try_fold(
+                current_types.first().unwrap().clone(),
+                |acc, x| {
+                    let corced_type = comparison_coercion(&acc, x);
+                    if let Some(corced_type) = corced_type {
+                        Ok(corced_type)
+                    } else {
+                        internal_err!("Coercion from {acc:?} to {x:?} failed.")
```

Comment:
We call unwrap_or previously so `select [1, true, null]` unexpectedly correct since true is castable to 1 in arrow-rs but not in datafusion.  `select [true, 1, null]` failed. It is better that we just return error if not coercible.